spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingHeaderNamesBuilder.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2018-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import org.springframework.util.Assert;
+
+/**
+ *
+ * Provides a convenient API for creating {@link DeadLetterPublishingRecoverer.HeaderNames}.
+ *
+ * @author Tomaz Fernandes
+ * @since 2.7
+ * @see DeadLetterPublishingRecoverer.HeaderNames
+ *
+ */
+public class DeadLetterPublishingHeaderNamesBuilder {
+
+	private final Original original = new Original();
+	private final Exception exception = new Exception();
+
+	public static DeadLetterPublishingHeaderNamesBuilder.Original original() {
+		return new DeadLetterPublishingHeaderNamesBuilder().original;
+	}
+
+	public class Original {
```

Comment:
Ok for the javadocs. As for the static class, I actually rely on this class not being static so that it can access the Builder's properties. This way I can provide a cleaner API without having to create too much of an infrastructure. So if I change this to static I'll probably have to change to a probably less clean API.

DYT it's a problem to leave it as it is?

Reply:
No; it's not a problem, but it is confusing to have these effectively duplicated in the DLPR.

Reply:
Yeah, I suppose it is. The main difference is that in the builder the fields are not final and it has the not null assertions in the build methods, while the HeaderNames class only has the final fields (no getters or setters or anything else).

I can try to come up with something less confusing, I'll look into that.

Reply:
Don't spend too much time; I can live with it now that I know the rationale.